File: java/com/google/copybara/folder/FolderOrigin.java
/*
 * Copyright (C) 2016 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.copybara.folder;

import static com.google.copybara.exception.ValidationException.checkCondition;

import com.google.common.base.Preconditions;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableListMultimap;
import com.google.common.collect.ImmutableSet;
import com.google.common.collect.ImmutableSetMultimap;
import com.google.copybara.Change;
import com.google.copybara.Origin;
import com.google.copybara.authoring.Author;
import com.google.copybara.authoring.Authoring;
import com.google.copybara.exception.RepoException;
import com.google.copybara.exception.ValidationException;
import com.google.copybara.util.AbsoluteSymlinksNotAllowed;
import com.google.copybara.util.FileUtil;
import com.google.copybara.util.FileUtil.CopySymlinkStrategy;
import com.google.copybara.util.Glob;
import java.io.IOException;
import java.nio.file.FileSystem;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.attribute.PosixFilePermission;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import javax.annotation.Nullable;

/**
 * Use a folder as the input for the migration.
 */
public class FolderOrigin implements Origin<FolderRevision> {

  private static final String LABEL_NAME = "FolderOrigin-RevId";
  private static final ImmutableSet<PosixFilePermission> FILE_PERMISSIONS =
      ImmutableSet.of(PosixFilePermission.OWNER_READ, PosixFilePermission.OWNER_WRITE);

  private final FileSystem fs;
  private final Author author;
  private final String message;
  private final Path cwd;
  private final CopySymlinkStrategy copySymlinkStrategy;

  FolderOrigin(FileSystem fs, Author author, String message, Path cwd,
      boolean materializeOutsideSymlinks, boolean ignoreInvalidSymlinks) {
    this.fs = Preconditions.checkNotNull(fs);
    this.author = author;
    this.message = message;
    this.cwd = Preconditions.checkNotNull(cwd);
    this.copySymlinkStrategy = ignoreInvalidSymlinks
            ? CopySymlinkStrategy.IGNORE_INVALID_SYMLINKS
            : materializeOutsideSymlinks
                ? CopySymlinkStrategy.MATERIALIZE_OUTSIDE_SYMLINKS
                : CopySymlinkStrategy.FAIL_OUTSIDE_SYMLINKS;
  }

  @Override
  public FolderRevision resolve(@Nullable String reference) throws ValidationException {
    checkCondition(reference != null, ""
        + "A path is expected as reference in the command line. Invoke copybara as:\n"
        + "    copybara copy.bara.sky workflow_name ORIGIN_FOLDER");
    Path path = fs.getPath(reference);
    if (!path.isAbsolute()) {
      path = cwd.resolve(path);
    }
    checkCondition(Files.exists(path), "%s folder doesn't exist", path);
    checkCondition(Files.isDirectory(path), "%s is not a folder", path);
    checkCondition(Files.isReadable(path) && Files.isExecutable(path), "%s is not readable/executable", path);

    return new FolderRevision(path, ZonedDateTime.now(ZoneId.systemDefault()));
  }

  @Override
  public Reader<FolderRevision> newReader(Glob originFiles, Authoring authoring)
      throws ValidationException {
    return new Reader<FolderRevision>() {
      @Override
      public void checkout(FolderRevision ref, Path workdir)
          throws RepoException, ValidationException {
        try {
          FileUtil.copyFilesRecursively(ref.path, workdir, copySymlinkStrategy, originFiles);
          FileUtil.addPermissionsAllRecursively(workdir, FILE_PERMISSIONS);
        } catch (AbsoluteSymlinksNotAllowed e) {
          throw new ValidationException(
              String.format("Cannot copy files into the workdir: Some"
                        + " symlinks refer to locations outside of the folder and"
                        + " 'materialize_outside_symlinks' config option was not used:\n"
                        + "  %s -> %s\n", e.getSymlink(), e.getDestinationFile()));
        } catch (IOException e) {
          throw new RepoException(
              String.format(
                  "Cannot copy files into the workdir:\n"
                      + "  origin folder: %s\n"
                      + "  workdir: %s",
                  ref.path, workdir),
              e);
        }
      }

      @Override
      public ChangesResponse<FolderRevision> changes(
          @Nullable FolderRevision fromRef, FolderRevision toRef) throws RepoException {
        // Ignore fromRef since a folder doesn't have history of changes
        return ChangesResponse.forChanges(ImmutableList.of(change(toRef)));
      }

      @Override
      public boolean supportsHistory() {
        return false;
      }

      @Override
      public Change<FolderRevision> change(FolderRevision ref) throws RepoException {
        return new Change<>(ref, author, message, ref.readTimestamp(), ImmutableListMultimap.of());
      }

      @Override
      public void visitChanges(FolderRevision start, ChangesVisitor visitor) throws RepoException {
        visitor.visit(change(start));
      }
    };
  }

  @Override
  public String getLabelName() {
    return LABEL_NAME;
  }

  @Override
  public String getType() {
    return "folder.origin";
  }

  @Override
  public ImmutableSetMultimap<String, String> describe(Glob destinationFiles) {
    return ImmutableSetMultimap.of("type", getType());
  }
}

File: src/Copybara.Core/Folder/FolderOrigin.cs
/*
 * Copyright (C) 2016 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System.Collections.Immutable;
using Copybara.Authoring;
using Copybara.Common;
using Copybara.Exceptions;
using Copybara.Revision;
using Copybara.Util;

namespace Copybara.Folder;

/// <summary>Use a folder as the input for the migration.</summary>
public class FolderOrigin : IOrigin<FolderRevision>
{
    private const string LabelName = "FolderOrigin-RevId";

    private const UnixFileMode FilePermissions = UnixFileMode.UserRead | UnixFileMode.UserWrite;

    private readonly string _fileSystemRoot;
    private readonly Author _author;
    private readonly string _message;
    private readonly string _cwd;
    private readonly FileUtil.CopySymlinkStrategy _copySymlinkStrategy;
    private readonly string? _version;

    internal FolderOrigin(
        string fileSystemRoot,
        Author author,
        string message,
        string cwd,
        FileUtil.CopySymlinkStrategy copySymlinkStrategy,
        string? version)
    {
        _fileSystemRoot = Preconditions.CheckNotNull(fileSystemRoot);
        _author = author;
        _message = message;
        _cwd = Preconditions.CheckNotNull(cwd);
        _copySymlinkStrategy = Preconditions.CheckNotNull(copySymlinkStrategy);
        _version = version;
    }

    public FolderRevision Resolve(string? reference)
    {
        ValidationException.CheckCondition(
            reference != null,
            "A path is expected as reference in the command line. Invoke copybara as:\n"
                + "    copybara copy.bara.sky workflow_name ORIGIN_FOLDER");
        string path = reference!;
        if (!Path.IsPathRooted(path))
        {
            path = Path.GetFullPath(Path.Combine(_cwd, path));
        }
        ValidationException.CheckCondition(
            Directory.Exists(path) || File.Exists(path), "%s folder doesn't exist", path);
        ValidationException.CheckCondition(Directory.Exists(path), "%s is not a folder", path);

        return new FolderRevision(path, DateTimeOffset.Now, _version);
    }

    public IOrigin<FolderRevision>.IReader<FolderRevision> NewReader(
        Glob originFiles, Authoring.Authoring authoring) =>
        new ReaderImpl(this, originFiles);

    public string GetLabelName() => LabelName;

    public string GetTypeName() => "folder.origin";

    public ImmutableListMultimap<string, string> Describe(Glob? originFiles)
    {
        var builder = ImmutableListMultimap<string, string>.CreateBuilder();
        builder.Put("type", GetTypeName());
        return builder.Build();
    }

    private sealed class ReaderImpl : IOrigin<FolderRevision>.IReader<FolderRevision>
    {
        private readonly FolderOrigin _origin;
        private readonly Glob _originFiles;

        public ReaderImpl(FolderOrigin origin, Glob originFiles)
        {
            _origin = origin;
            _originFiles = originFiles;
        }

        public void Checkout(FolderRevision reference, string checkoutDir)
        {
            try
            {
                FileUtil.CopyFilesRecursively(
                    reference.Path, checkoutDir, _origin._copySymlinkStrategy, _originFiles);
                FileUtil.AddPermissionsAllRecursively(checkoutDir, FilePermissions);
            }
            catch (SymlinkException e)
            {
                throw new ValidationException(
                    "Cannot copy files into the workdir: " + e.Message, e);
            }
            catch (IOException e)
            {
                throw new RepoException(
                    "Cannot copy files into the workdir:\n"
                        + $"  origin folder: {reference.Path}\n"
                        + $"  workdir: {checkoutDir}",
                    e);
            }
        }

        public Origin.ChangesResponse<FolderRevision> Changes(
            FolderRevision? fromRef, FolderRevision toRef)
        {
            // Ignore fromRef since a folder doesn't have history of changes
            return Origin.ChangesResponse<FolderRevision>.ForChanges(
                ImmutableArray.Create(Change(toRef)));
        }

        public bool SupportsHistory() => false;

        public Change<FolderRevision> Change(FolderRevision reference) =>
            new(
                reference,
                _origin._author,
                _origin._message,
                reference.ReadTimestamp() ?? DateTimeOffset.Now,
                ImmutableListMultimap<string, string>.Empty);

        public void VisitChanges(FolderRevision? start, IChangesVisitor visitor)
        {
            FolderRevision reference = start!;
            var change = new Change<IRevision>(
                reference,
                _origin._author,
                _origin._message,
                reference.ReadTimestamp() ?? DateTimeOffset.Now,
                ImmutableListMultimap<string, string>.Empty);
            visitor.Visit(change);
        }
    }
}
